control_plane/attachment_service/src/service.rs

```diff
@@ -505,7 +555,18 @@ impl Service {
             // Block the [`Service::startup_complete`] barrier until we're done
             let _completion = startup_completion;
 
-            startup_reconcile_this.startup_reconcile().await
+            // Block shutdown until we're done (we must respect self.cancel)
+            if let Ok(_gate) = startup_reconcile_this.gate.enter() {
+                startup_reconcile_this.startup_reconcile().await
+            }
+        });
+
+        let background_reconcile_this = this.clone();
+        tokio::task::spawn(async move {
+            // Block shutdown until we're done (we must respect self.cancel)
+            if let Ok(_gate) = background_reconcile_this.gate.enter() {
+                background_reconcile_this.background_reconcile().await
+            }
```

Comment:
Could you move this in the `spawn` body above (drop the gate stuff & move _completion above the call to `startup_reconcile)? This way you don't have to wait for startup completion explicitly.
I think it might read a bit nicer.

Reply:
Refactored this.